File: pset6/credit.py
"""
    product() return total of the product of the number passed in. eg: 14 = 1 + 4 = 5
"""
def product(n):
    total = 0

    for i in str(n):
        total += int(i)

    return total


"""
    checksum() checks a card number and determines if it is authentic or not. returns boolean
"""
def checksum(n):
    total = 0
    # reverse card number for better readability
    n = n[::-1]

    # grab and double every odd digit
    odd_digits = [int(i) * 2 for i in n[1::2]]
    # grab every even digit
    even_digits = [int(i) for i in n[::2]]

    # add the products of the odd indexed numbers together
    for digit in odd_digits:
        if digit >= 10:
            total += product(digit)
        else:
            total += digit

    # add sum of even digits
    total += sum(even_digits)

    # determine if valid
    if total % 10 == 0:
        return True
    else:
        return False


while True:
    cardno = input("Number: ")
    # check card number is a valid integer (without changing it to one)
    try:
        int(cardno)
    except  ValueError:
        print("Card number must be an integer!")
        continue
    else:
        break

# get length of number
l = len(cardno)

card = "INVALID"

# check card attributes and assign card type string if checksum passes 
if (l == 13 or l == 16) and cardno[0] == '4':
    if checksum(cardno):
        card = "VISA"
elif l == 15 and cardno[:2] in ['34', '37']:
    if checksum(cardno):
        card = "AMEX"
elif l == 16 and cardno[:2] in ['51', '52', '53', '54', '55']:
    if checksum(cardno):
        card = "MASTERCARD"

print(card)

File: pset1/credit.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <stdlib.h>
/*
    determines if a credit card number is valid or not and which type of credit card it belongs to if so.
*/

int main (void){
    
    // get user input
    long long card_no;
    printf("Number: ");
    while(1){
        card_no = get_long_long();
        if (card_no > 0){
            break;
        } // else
        printf("Retry: ");
    }
    

    /* determine cardtype */
    char card_str[20]; 
    sprintf(card_str, "%lld", card_no);
    int len = strlen(card_str);
    
    // get first two numbers of card (adjust for ascii)
    int x = card_str[0] - 48;
    int y = card_str[1] - 48;

    char type[12];
    if ((len == 13 || len == 16) && x == 4){
        strcpy(type, "VISA");
    } else if (len == 16 && x == 5 && (y >= 1 && y <= 5)){
        strcpy(type, "MASTERCARD");
    } else if (len == 15 && x == 3 && (y == 4 || y == 7)){
        strcpy(type, "AMEX");
    } else {
        printf("INVALID\n");
        return 0;
    }
   
   
    /* checksum */
    bool odd = true;
    int checksum = 0;
    
    // collecting both odds and even indexed digits (working from the back of card_no) in one loop.
        // the odd indexed digits are added to the checksum unmodified
        // the even indexed digits will be multiplied by two and the digits of that sum added to the checksum;  
    while (card_no){
        int num = card_no % 10;
        card_no /= 10;
    
        if(odd){
            odd = false;
        } else {
            num *= 2;
            // get product of digits if greater than two numbers.
            if (num >= 10) {
                int temp = num % 10;
                temp += num /= 10;
                num = temp;
            }
            odd = true;
        }
        
        checksum += num;
    }
    
    // validate
    if (checksum % 10 != 0) {
        printf("INVALID\n");
    } else {
        printf("%s\n", type);
    }
    return 0;

}